hazelcast/src/hazelcast/client/spi/impl/ClientInvocation.cpp

```diff
@@ -0,0 +1,398 @@
+                    return boost::shared_ptr<ClientInvocation>(
+                            new ClientInvocation(clientContext, clientMessage, objectName));
+                }
+
+                boost::shared_ptr<connection::Connection> ClientInvocation::getSendConnection() {
+                    return sendConnection;
+                }
+
+                void
+                ClientInvocation::setSendConnection(const boost::shared_ptr<connection::Connection> &sendConnection) {
+                    ClientInvocation::sendConnection = sendConnection;
+                }
+
+                void ClientInvocation::notify(const boost::shared_ptr<protocol::ClientMessage> &clientMessage) {
+                    if (clientMessage.get() == NULL) {
+                        throw exception::IllegalArgumentException("response can't be null");
+                    }
+                    clientInvocationFuture->complete(clientMessage);
+                }
+
+                const boost::shared_ptr<protocol::ClientMessage> &ClientInvocation::getClientMessage() const {
+                    return clientMessage;
+                }
+
+                bool ClientInvocation::shouldBypassHeartbeatCheck() const {
+                    return bypassHeartbeatCheck;
+                }
+
+                void ClientInvocation::setBypassHeartbeatCheck(bool bypassHeartbeatCheck) {
+                    ClientInvocation::bypassHeartbeatCheck = bypassHeartbeatCheck;
+                }
+
+                const boost::shared_ptr<EventHandler<protocol::ClientMessage> > &
+                ClientInvocation::getEventHandler() const {
+                    return eventHandler;
+                }
+
+                void ClientInvocation::setEventHandler(
+                        const boost::shared_ptr<EventHandler<protocol::ClientMessage> > &eventHandler) {
+                    ClientInvocation::eventHandler = eventHandler;
+                }
+
+                void ClientInvocation::execute() {
+                    if (invokeCount < MAX_FAST_INVOCATION_COUNT) {
+                        // fast retry for the first few invocations
+                        executionService.execute(boost::shared_ptr<util::Runnable>(shared_from_this()));
+                    } else {
+                        // progressive retry delay
+                        long delayMillis = util::min<int64_t>(1 << ((int64_t) invokeCount - MAX_FAST_INVOCATION_COUNT),
+                                                              retryPauseMillis);
+                        executionService.schedule(shared_from_this(), delayMillis);
+                    }
+                }
+
+                const std::string ClientInvocation::getName() const {
+                    std::ostringstream out;
+                    out << *this;
+                    return out.str();
+                }
+
+                ClientInvocation::~ClientInvocation() {
```

Comment:
Unused destructor. Can be deleted.

Reply:
Since it is used in boost::shared_ptr we need virtual destructor.